### PR TITLE
feat: auth plugins access to dist tag info

### LINKED ISSUE #1686: dist tag based permissions
**Is your feature request related to a problem?**
I would like to allow publishing for authenticated users to publish to "development" dist tag but would like to restrict "latest" tag to be permitted by ci build servers.

**Describe the solution you'd like**
I would like to expose dist tag information to plugins, this way a plugin could implement the authentication logic on weather or not grant permission to the user depending on what tag they are publishing to.

**Describe alternatives you've considered**
I have looked at adding it to the core package authentication configuration. It seems as though it may be confusing to configure. "@myorg/***/@development". It seems like it wouldn't be common so might be better as a plugin, but the data needs to be passed along to the auth plugin api.

### PR DESCRIPTION
**Pull request by g3hxqe-contrib**
**Description:**
I would like to expose dist tag information to plugins, this way a plugin could implement the authentication logic on weather or not grant permission to the user depending on what tag they are publishing to.

Resolves #1686

### CODE REVIEW

**Comment by g3hxqe-contrib** [comment]
Example plugin, very much MVP: https://github.com/g3hxqe-contrib/verdaccio-tagauth

**Comment by g3hxqe-contrib** [comment]
corresponding type updates: https://github.com/verdaccio/monorepo/pull/318

**Comment by juanpicado** [comment]
Thanks @g3hxqe-contrib , I'll review this in the next few days. Also cc: @verdaccio/collaborators would be appreciated.

File: src/api/middleware.ts [hunk 1]
```diff
@@ -104,9 +104,10 @@ export function allow(auth: IAuth): Function {
       const packageName = req.params.scope ? `@${req.params.scope}/${req.params.package}` : req.params.package;
       const packageVersion = req.params.filename ? getVersionFromTarball(req.params.filename) : undefined;
       const remote: RemoteUser = req.remote_user;
+      const tag = req.body['dist-tags'] ? Object.keys(req.body['dist-tags']).pop() : 'latest';
```
Comment (anikethsaha) [inline src/api/middleware.ts]: Just one thing, I think it will make more sense to change `tag` -> `dist_tag` ?

Comment (juanpicado) [reply src/api/middleware.ts]: My opinion here is, hardcode `latest` is not a good idea, `latest` is the default tag but actually can be removed from any package manually. I've seen scenarios here where people uses other names by default and `latest` is missing completely.  I'm still checking this anyway.

**Pull request closed without merging**